### DISCUSSION TITLE
Enabling orography filtering for individual interpolation methods

### DISCUSSION THREAD

**Discussion by heu3becteh** [dtcenter/METplus#2356] [Configuration] [answered]
Hello!

For our verification we need to use various interpolation methods. In particular, for mountainous regions it is preferable to use methods that take into account differences in orography.
I asked about possible ways to find the nearest grid point taking into account the orography in [another thread](https://github.com/dtcenter/METplus/discussions/2186), at the moment `GEOG_MATCH` seems to be the best solution.

I have configured **PointStat** this way:

```
interp = {
   vld_thresh = 1.0;
   shape      = SQUARE;

   type = [
      {
         method = NEAREST;
         width  = 1;
      },
      {
         method = BILIN;
         width  = 2;
      },
      {
         method = GEOG_MATCH;
         width  = 5;
      },
      {
         method = UW_MEAN;
         width  = 5;
      }
   ];
}

land_mask = {
   flag      = TRUE;
   file_name = [ "${METPLUS_land_mask}" ];
   field     = { name = "LAND"; level = "L0"; }
   regrid    = { method = NEAREST; width = 1; }
   thresh    = eq1;
}

topo_mask = {
   flag               = TRUE;
   file_name          = [ "${METPLUS_topo_mask}" ];
   field              = { name = "DIST"; level = "L0"; }
   regrid             = { method = NEAREST; width = 1; }
   use_obs_thresh     = ge-100&&le100;
   interp_fcst_thresh = ge-50&&le50;
}
```

The problem with this approach is that `topo_mask` is used for all interpolation methods, while I would like to enable it only for `GEOG_MATCH`, which results (generally) in the selection of the optimal grid point rather than the removal of the forecast-observation pair.
Even with `interp_fcst_thresh` disabled, the `UW_MEAN` interpolation method loses a significant portion of data. `interp_fcst_thresh` filters out the vast majority of `UW_MEAN` pairs.

It would be very convenient to enable orography filtering only for selected methods.

In the [user manual](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html#topo-mask), I saw the setting to enable/disable orography filtering for individual fields:
> The “topo_mask.flag” entry may be set separately in each “obs.field” entry.

Is this possible for individual interpolation methods?
Is it necessary to do separate runs for different interpolation methods with orography filtering enabled and disabled?

Best regards,
Alexander

**Comment by JohnHalleyGotway** [comment, marked as answer]
Alexander, I see that you have a question about configuring Point-Stat to specify how the `topo_mask` setting is applied. You'd like to set `topo_mask.flag = TRUE` but only when the interpolation `method = GEOG_MATCH`, and set `topo_mask.flag = FALSE` for all other interpolation methods.

The simple answer is no, `topo_mask.flag` cannot be set separately for each interpolation entry. The reason is that topo_mask applies filtering logic to the point observations. It actually discards observations which don't meet the criteria. So the number of fcst/obs matched pairs will differ with it on/off. Supporting turning it on/off for each interpolation method would make the processing logic overly complex and difficult to maintain.

However, as indicated in the [documentation you referenced](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html#topo-mask):
```
The “topo_mask.flag” entry may be set separately in each “obs.field” entry.
```
I see two reasonable options available to you, and it's up to you to decide which you prefer.

1. Run Point-Stat twice, once with topo_mask disabled for the NEAREST, BILIN, and UW_MEAN interpolation methods and a second time with topo_mask enabled for the GEOG_MATCH method. Running it twice with the same input files will cause it to overwrite its output. So be sure to set the [POINT_STAT_OUTPUT_PREFIX](https://metplus.readthedocs.io/en/latest/Users_Guide/glossary.html#term-POINT_STAT_OUTPUT_PREFIX) METplus config option to make the output file names unique for each run.

2. The second option is running Point-Stat only once, but listing the same field (or fields) twice. The first time, you'd set the [POINT_STAT_OBS_VAR1_OPTIONS](https://metplus.readthedocs.io/en/latest/Users_Guide/glossary.html#term-OBS_POINT_STAT_VAR-n-_OPTIONS) METplus config option to specify 3 interpolation methods and `topo_mask.flag = FALSE`. Then you'd set [POINT_STAT_OBS_VAR2_OPTIONS](https://metplus.readthedocs.io/en/latest/Users_Guide/glossary.html#term-OBS_POINT_STAT_VAR-n-_OPTIONS) to specify the GEOG_MATCH interpolation method and `topo_mask.flag = TRUE`.

Which of those two approaches sounds more appealing to you?

**Comment by heu3becteh** [comment]
Thank you for your reply!

> I see two reasonable options available to you, and it's up to you to decide which you prefer.
> 
>     1. Run Point-Stat twice, once with topo_mask disabled for the NEAREST, BILIN, and UW_MEAN interpolation methods and a second time with topo_mask enabled for the GEOG_MATCH method. Running it twice with the same input files will cause it to overwrite its output. So be sure to set the [POINT_STAT_OUTPUT_PREFIX](https://metplus.readthedocs.io/en/latest/Users_Guide/glossary.html#term-POINT_STAT_OUTPUT_PREFIX) METplus config option to make the output file names unique for each run.
> 
>     2. The second option is running Point-Stat only once, but listing the same field (or fields) twice. The first time, you'd set the [POINT_STAT_OBS_VAR1_OPTIONS](https://metplus.readthedocs.io/en/latest/Users_Guide/glossary.html#term-OBS_POINT_STAT_VAR-n-_OPTIONS) METplus config option to specify 3 interpolation methods and `topo_mask.flag = FALSE`. Then you'd set [POINT_STAT_OBS_VAR2_OPTIONS](https://metplus.readthedocs.io/en/latest/Users_Guide/glossary.html#term-OBS_POINT_STAT_VAR-n-_OPTIONS) to specify the GEOG_MATCH interpolation method and `topo_mask.flag = TRUE`.
> 
> 
> Which of those two approaches sounds more appealing to you?

I think about running Point-Stat twice.
It is not much of a problem to set more branches for output using `POINT_STAT_OUTPUT_TEMPLATE` in METplus config-file, there are already a batch of directories for different configurations.
What concerns me more is that calculation time will increase, I will check how much it will.

We are using `mpr_column` and `mpr_thresh` in our PointStatConfig wrapper, I assume it is problematic to do in [POINT_STAT_OBS_VAR-n_OPTIONS](https://metplus.readthedocs.io/en/latest/Users_Guide/glossary.html#term-OBS_POINT_STAT_VAR-n-_OPTIONS) and you can not place `interp` methods into `field` in wrapper.

Best regards,
Alexander

**Reply by JohnHalleyGotway** [reply]
Hi Alexander. Just wanted to check back in on this discussion from last September.

Sorry for losing track of this one. It's easy to miss notifications on discussions that have already been marked as being answered. That's why we typically "lock" them to prevent future posts.

Do any questions remain, or is your configuration working well now?

**Reply by heu3becteh** [reply]
Hi!

The configuration is working as intended (PointStat is running twice, with two different wrappers), but our research has shown that different grid-to-point interpolation methods are needed for different fields. At the moment we cannot afford to run all the interpolation methods for each field, the system starts to work slowly and takes up a lot of disk space. Therefore, we will have to create separate wrappers for individual variables, as well as a separate one for orographic filtering.
At the moment, PointStat execution time is less of an issue to us than disk space. So we'll probably run PointStat multiple times with different wrappers. There are general questions about verification workflow, but this is a separate topic, I hope to organize questions about this in the near future in a separate thread.

Best regards,
Alexander

**Reply by JohnHalleyGotway** [reply]
Alexander, it is possible in the Point-Stat config file to specify the grid-to-point interpolation method separately for each verification task. Let me offer some sample config settings to illustrate. Let's say you're verifying a forecast file that contains 500mb temperature (TMP/P500), total cloud amount (TCDC), and 6-hour accumulated precip (APCP/A6).

1. Runs all message types and interpolation methods for all fields:
```
fcst = {
   field = [
      { name="TMP"; level="P500"; },
      { name="TCDC"; level="Z0"; },
      { name="APCP"; level="A6"; }
   ];
}
obs = {
   field = [
      { name="TMP"; level="P500"; },
      { name="TCDC"; level="Z0"; },
      { name="APCP"; level="A6"; }
   ];
}
message_type = [ "ADPSFC", "ADPUPA" ];
interp = {
   vld_thresh = 1.0;
   shape      = SQUARE;
   type = [ { method = NEAREST; width = 1; },
            { method = BILIN;   width = 2; },
            { method = BUDGET;  width = 2; } ];
}
```
This would produce output for 18 verification tasks: 3 fields x 2 message types x 3 interpolation methods.

2. Specify a single message type and interpolation method for each field:
```
fcst = {
   field = [
      { name="TMP"; level="P500"; },
      { name="TCDC"; level="Z0"; },
      { name="APCP"; level="A6"; }
   ];
}
obs = {
   field = [
      { name="TMP"; level="P500";
        message_type = "ADPUPA";
        interp = { type = [ { method = BILIN; width = 2; } ]; }},
      { name="TCDC"; level="Z0";
        message_type = "ADPSFC";
        interp = { type = [ { method = NEAREST; width = 1; } ]; }},
      { name="APCP"; level="A6";
        message_type = "ADPSFC";
        interp = { type = [ { method = BUDGET; width = 2; } ]; }}
   ];
}
```
This would produce output for only 3 verification tasks.
I'm hoping I got all the commas and braces in the right spot. But I just wanted to illustrate that these things can set separately for each verification task. That is exactly what is meant by [this comment](https://github.com/dtcenter/MET/blob/df8fa78ce0373615937684f3d79768e1fae0d80f/data/config/PointStatConfig_default#L95) in the Point-Stat config files:
`// May be set separately in each "obs.field" entry`

The question is how to go about configuring these settings through the METplus configuration file. And for that, I'll request that @georgemccabe chime in to see if he has any recommendations.
